### PR TITLE
Scan exec blog post

### PR DESCRIPTION
**Pull request by keith-turner**

### CODE REVIEW

File: _posts/blog/2019-09-30-scan-executors.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
+---
+title: "How Fluo Leveraged Scan Executors"
+date: 2019-09-30 11:50:00 +0000
+author: Keith Turner
+---
+
+Accumulo 2.0 introduced [Scan Executors][1] giving control over processing of
+scans in Accumulo tablet servers. Fluo has a good use case for scan executors,
```
Comment (ctubbsii) [inline _posts/blog/2019-09-30-scan-executors.md]:
```suggestion
scans in Accumulo tablet servers. Fluo has a good use case for scan executors:
```

File: _posts/blog/2019-09-30-scan-executors.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+notification scans.  Fluo worker processes continually scan for notifications
+that indicate there is work to do. All workers continually scanning for
+notifications can place a lot of load on Accumulo tablet servers which could
+negatively impact transactions.  The new scan executor feature provides a way
+to limit this load.
+
+Fluo utilizes this feature by [setting scan hints][2] for notification scans
+indicating `scan_type=fluo-ntfy`.  These hints are passed to Accumulo tablet
+servers and are ignored by default. For these scan types Accumulo could be
```
Comment (ctubbsii) [inline _posts/blog/2019-09-30-scan-executors.md]:
```suggestion
servers and are ignored by default. For these scan types, Accumulo could be
```

File: _posts/blog/2019-09-30-scan-executors.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+configured to either send them to special thread pool and/or prioritize them
```
Comment (ctubbsii) [inline _posts/blog/2019-09-30-scan-executors.md]:
```suggestion
configured to either send them to a special thread pool and/or prioritize them
```

File: _posts/blog/2019-09-30-scan-executors.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+differently within a thread pool.  Below is an example of Accumulo shell
+commands that setup a special executor for notification scans.
```
Comment (ctubbsii) [inline _posts/blog/2019-09-30-scan-executors.md]:
```suggestion
commands that set up a special executor for notification scans.
```

File: _posts/blog/2019-09-30-scan-executors.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+```
+config -s tserver.scan.executors.fnotify.threads=1
+config -t fluo_table -s table.scan.dispatcher=org.apache.accumulo.core.spi.scan.SimpleScanDispatcher
+config -t fluo_table -s table.scan.dispatcher.opts.executor.fluo-ntfy=fnotify
+```
+
+The system setting `tserver.scan.executors.fnotify.threads=1` creates a single
```
Comment (ctubbsii) [inline _posts/blog/2019-09-30-scan-executors.md]:
```suggestion
The system setting `tserver.scan.executors.fnotify.threads=1` creates a single-threaded
```

File: _posts/blog/2019-09-30-scan-executors.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+threaded scan executor in each tablet server named `fnotify`. The two per table
```
Comment (ctubbsii) [inline _posts/blog/2019-09-30-scan-executors.md]:
```suggestion
scan executor in each tablet server named `fnotify`. The two per-table
```

File: _posts/blog/2019-09-30-scan-executors.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+settings configure a scan dispatcher (the SimpleScanDispatcher is built into
```
Comment (ctubbsii) [inline _posts/blog/2019-09-30-scan-executors.md]:
```suggestion
settings configure a scan dispatcher (the `SimpleScanDispatcher` is built into
```

File: _posts/blog/2019-09-30-scan-executors.md [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+Accumulo) on the fluo table.  The scan dispatcher is configured such that when
```
Comment (ctubbsii) [inline _posts/blog/2019-09-30-scan-executors.md]:
```suggestion
Accumulo) on the table `fluo_table`.  The scan dispatcher is configured such that when
```

File: _posts/blog/2019-09-30-scan-executors.md [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+a scan hint of `scan_type=fluo-ntfy` is seen it runs on the executor `fnotify`.
```
Comment (ctubbsii) [inline _posts/blog/2019-09-30-scan-executors.md]:
```suggestion
a scan hint of `scan_type=fluo-ntfy` is seen, it runs on the executor `fnotify`.
```

File: _posts/blog/2019-09-30-scan-executors.md [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+All other scans will run on the default executor. This has the effect running
+all notification scans on a single dedicated thread in each tablet server.
+
+The above setting were tested in a scenario where 20 Fluo worker were run
+against a single tablet server with 20 tablets.  The Fluo stress test was run
+with a low ingest rate, resulting in continual notification scanning by the 20
+workers.  While the test was running, jstack and top were used to inspect the
```
Comment (ctubbsii) [inline _posts/blog/2019-09-30-scan-executors.md]:
```suggestion
workers.  While the test was running, `jstack` and `top` were used to inspect the
```

File: _posts/blog/2019-09-30-scan-executors.md [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+tablet server. This inspection revealed that notification scans were all
+running in single thread which was using 100% of a single core.  This left all
```
Comment (ctubbsii) [inline _posts/blog/2019-09-30-scan-executors.md]:
```suggestion
running in a single thread which was using 100% of a single core.  This left all
```

File: _posts/blog/2019-09-30-scan-executors.md [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+of the other cores free to process transactions.  Further testing to see how
+this impacts throughput is needed. Observing the worker debug logs, all of them
+seemed to complete notification scans quickly finding new work.
```
Comment (ctubbsii) [inline _posts/blog/2019-09-30-scan-executors.md]:
```suggestion
seemed to complete notification scans, quickly finding new work.
```

File: _posts/blog/2019-09-30-scan-executors.md [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+
+Fluo took a descriptive approach to using scan hints, where it described what
```
Comment (ctubbsii) [inline _posts/blog/2019-09-30-scan-executors.md]:
```suggestion
Fluo took a descriptive approach to using scan hints, where it described to Accumulo
```

File: _posts/blog/2019-09-30-scan-executors.md [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+type of scan was running to Accumulo.  However, Fluo does not care what if
```
Comment (ctubbsii) [inline _posts/blog/2019-09-30-scan-executors.md]:
```suggestion
the type of scan it was running.  However, Fluo does not care what, if
```

File: _posts/blog/2019-09-30-scan-executors.md [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+anything Accumulo does with that information.  This allows administrators to
```
Comment (ctubbsii) [inline _posts/blog/2019-09-30-scan-executors.md]:
```suggestion
anything, Accumulo does with that information.  This allows administrators to
```

File: _posts/blog/2019-09-30-scan-executors.md [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+configure Accumulo in many different ways to handle notification scans, without
+any changes to Fluo.
+
+For my first pass at using scan executors I tried a prescriptive approach. I
```
Comment (ctubbsii) [inline _posts/blog/2019-09-30-scan-executors.md]:
```suggestion
For my first pass at using scan executors, I tried a prescriptive approach. I
```

File: _posts/blog/2019-09-30-scan-executors.md [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+attempted to use scan hints to explictily name an executor for notification
+scans.  I realized this would require Fluo configuration to provide the name of
+the scan executor. Forcing a user to specify Accumulo and Fluo configuration
+was very cumbersome so I abandoned the prescriptive approach.  The descriptive
+approach I settled on its place is less cumbersome (it only requires Accumulo
```
Comment (ctubbsii) [inline _posts/blog/2019-09-30-scan-executors.md]:
```suggestion
approach I settled on in its place is less cumbersome (it only requires Accumulo
```

File: _posts/blog/2019-09-30-scan-executors.md [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+config) and more flexible (it supports executors and/or prioritization instead
+of only executors).
+
+At the time of this writing no released version of Fluo supports Accumulo 2.0.
```
Comment (ctubbsii) [inline _posts/blog/2019-09-30-scan-executors.md]:
```suggestion
At the time of this writing, no released version of Fluo supports Accumulo 2.0.
```

File: _posts/blog/2019-09-30-scan-executors.md [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+Once Fluo 1.3.0 is released with Accumulo 2.0, Hadoop 3.0, and Java 11 support
```
Comment (ctubbsii) [inline _posts/blog/2019-09-30-scan-executors.md]:
```suggestion
Once Fluo 1.3.0 is released with Accumulo 2.0, Hadoop 3.0, and Java 11 support,
```

**Review by ctubbsii — COMMENTED** [review]
Minor grammar cleanup. Otherwise, LGTM.

File: _posts/blog/2019-09-30-scan-executors.md [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+all notification scans on a single dedicated thread in each tablet server.
+
+The above setting were tested in a scenario where 20 Fluo worker were run
+against a single tablet server with 20 tablets.  The Fluo stress test was run
+with a low ingest rate, resulting in continual notification scanning by the 20
+workers.  While the test was running, jstack and top were used to inspect the
+tablet server. This inspection revealed that notification scans were all
+running in single thread which was using 100% of a single core.  This left all
+of the other cores free to process transactions.  Further testing to see how
+this impacts throughput is needed. Observing the worker debug logs, all of them
```
Comment (cjmctague) [inline _posts/blog/2019-09-30-scan-executors.md:42]: "All of them" -> "all them."

File: _posts/blog/2019-09-30-scan-executors.md [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+seemed to complete notification scans quickly finding new work.
+
+Fluo took a descriptive approach to using scan hints, where it described what
+type of scan was running to Accumulo.  However, Fluo does not care what if
+anything Accumulo does with that information.  This allows administrators to
+configure Accumulo in many different ways to handle notification scans, without
+any changes to Fluo.
+
+For my first pass at using scan executors I tried a prescriptive approach. I
+attempted to use scan hints to explictily name an executor for notification
```
Comment (cjmctague) [inline _posts/blog/2019-09-30-scan-executors.md]: "explictily" -> "explicitly"

File: _posts/blog/2019-09-30-scan-executors.md (hunk 20 above)
Comment (ctubbsii) [reply _posts/blog/2019-09-30-scan-executors.md:42]: "all of them" is correct.

File: _posts/blog/2019-09-30-scan-executors.md (hunk 21 above)
Comment (ctubbsii) [reply _posts/blog/2019-09-30-scan-executors.md]:
```suggestion
attempted to use scan hints to explicitly name an executor for notification
```

**Comment by ctubbsii** [comment]
@keith-turner I recently figured out how to batch lots of little suggestions into a single commit: you have to do it from the "Files Changed" tab.

Also, I saw that they recently added a multi-line select-and-comment feature to the reviews... but multi-line suggested changes don't work yet (it still only uses the last line selected). Hopefully, multi-line suggested changes will work soon.

**Review by ctubbsii — APPROVED** [review]

**Pull request merged**